decision_service/dsclientlib/ds_error_callback_fn.h

```diff
@@ -0,0 +1,28 @@
+#pragma once
+#include <mutex>
+
+namespace decision_service {
+  class api_status;
+
+  class error_callback_fn
+  {
+    public:
+      using error_fn = void(*)(const api_status&, void*);
```

Comment:
The way I imagine this being used by an API consumer is to send the API an opaque pointer along with the function to call back.  Generally this is a pointer to an object that the user will cast back into that object (since they know what they sent)  and then switch to c++ code to handle the error.

1) Since error_fn is going to be called back with the value passed in as void*, this order felt more natural to me.  However I am happy to change it if this seems odd.

2) We cannot carry the type information from the user into the background thread without needing to recompile the entire library.  Should be possible to loose the type information at the interface layer.   However that would mean the driver constructor will be recompiled by the client.  

---
In reply to: [187100996](https://github.com/JohnLangford/vowpal_wabbit/pull/1479#discussion_r187100996) [](ancestors = 187100996)

Reply:
Perhaps we should have both an alternate header-only type-checking interface and the void* interface?  That will help document what the void* is supposed to do and provides easy type checking for those wanting to compile?